### PR TITLE
handle tx finalization in controllers instead of provider-engine

### LINKED ISSUE #1001: Improve Tx Signature Flow
```
hookedSubp.handleRequest: eth_sendTransaction

  // basic validity check

  hookedSubp.validateTransaction
    // checks that sender is in accounts
    hookedSubp.validateSender

  // UI approval hook

  hookedSubp.approveTransaction
    metamaskController.newUnsignedTransaction
      metamaskController.enforceTxValidations
      txManager.addUnapprovedTransaction
        txProviderUtils.analyzeGasUsage
        txManager.txDidComplete
          txManager.addTx
            txManager.once(`${txMeta.id}:signed`)
            txManager.once(`${txMeta.id}:rejected`)
      // trigger popup
      metamaskController.opts.showUnapprovedTx

  (ui interaction)
    metamaskController.api.approveTransaction
      txManager.approveTransaction
        txManager.setTxStatusSigned(txId)
          txManager._setTxStatus
            txManager.emit(`${txMeta.id}:${status}`, txId)

  hookedSubp.checkApproval

  // tx signing hook

  hookedSubp.finalizeAndSubmitTx
    // bottle neck to ensure atomic nonce
    hookedSubp.nonceLock.take
      // set fallback gasPrice, nonce, etc
      hookedSubp.fillInTxExtras
      hookedSubp.signTransaction
        txManager.formatTxForSigining(txParams)
        keyringController.signTransaction(ethTx, address, txId)
        txManager.resolveSignedTransaction({tx, txId})
        txManager.emit(`${txParams.metamaskId}:signingComplete`)
      // sendRawTransaction
      hookedSubp.publishTransaction
```

Problems:

  * not persistence friendly
  * hard to share tx metadata between "approveTransaction" and "finalizeAndSubmitTx"

Possible Solutions:

 * unify hookedWallet <-> MetaMask hooks into a single hook
   * neg: have to reimplement nonceLock, fillInTxExtras
 * make hookedWallet wrap txs in its own meta obj
   * persistence is still and issue

### PR DESCRIPTION
**Pull request by kumavis**
involved copying a ton of finalization code out of provider-engine and into metamask : /
since we now broadcast the tx ourself, we dont have the issue of submitting txs from persistence

Fixes #1001

### CODE REVIEW

File: app/scripts/metamask-controller.js [hunk 1]
```diff
@@ -257,24 +239,29 @@ module.exports = class MetamaskController extends EventEmitter {
...
+        cb(null, rawTx)
+      }
+      function failHandler() {
+        removeHandlers()
+        cb(new Error('User denied message signature.'))
+      }
+      function removeHandlers() {
+        this.txManager.removeListener(`${txMeta.id}:submitted`, successHandler)
+        this.txManager.removeListener(`${txMeta.id}:rejected`, failHandler)
+      }
```
Comment (kumavis) [inline app/scripts/metamask-controller.js]: i reccomend adding a single method for complete/fail:
```
this.txManager.once(`${txMeta.id}:finished`, failHandler)
```
can pass the txId

File: app/scripts/metamask-controller.js [hunk 2]
```diff
@@ -257,24 +239,29 @@ module.exports = class MetamaskController extends EventEmitter {
...
+  newUnapprovedTransaction (txParams, cb) {
+    this.txManager.addUnapprovedTransaction(txParams, (err, txMeta) => {
+      if (err) return cb(err)
       this.sendUpdate()
-      this.opts.showUnapprovedTx(txParams, txData, onTxDoneCb)
+      this.opts.showUnapprovedTx(txMeta)
+      // listen for tx completion (success, fail)
+      this.txManager.once(`${txMeta.id}:submitted`, successHandler)
+      this.txManager.once(`${txMeta.id}:rejected`, failHandler)
+      function successHandler(rawTx) {
```
Comment (kumavis) [inline app/scripts/metamask-controller.js]: actually we dont care about the rawTx what we need to return is the txHash

File: app/scripts/transaction-manager.js [hunk 3]
```diff
@@ -15,11 +14,14 @@ module.exports = class TransactionManager extends EventEmitter {
     this.txHistoryLimit = opts.txHistoryLimit
     this.getSelectedAccount = opts.getSelectedAccount
     this.provider = opts.provider
+    this.query = opts.query
```
Comment (kumavis) [inline app/scripts/transaction-manager.js]: remove

File: app/scripts/transaction-manager.js [hunk 4]
```diff
@@ -144,46 +145,76 @@ module.exports = class TransactionManager extends EventEmitter {
...
+    // only allow one tx at a time for atomic nonce usage
+    self.nonceLock.take(() => {
+      // begin signature process
+      async.waterfall([
+        (cb) => self.fillInTxParams(txId, cb),
+        (cb) => self.signTransaction(txId, cb),
+        (rawTx, cb) => self.publishTransaction(txId, rawTx, cb),
+      ], (err) => {
+        self.nonceLock.leave()
+        // TODO: move tx to error state
```
Comment (kumavis) [inline app/scripts/transaction-manager.js]: move tx to error state ?

File: app/scripts/transaction-manager.js [hunk 5]
```diff
@@ -144,46 +145,76 @@ module.exports = class TransactionManager extends EventEmitter {
...
+      cb(null, ethUtil.bufferToHex(ethTx.serialize()))
+    }).catch((err) => {
+      cb(err)
+    })
+  }
+
+  publishTransaction (txId, rawTx, cb) {
+    this.txProviderUtils.publishTransaction(rawTx, (err) => {
+      if (err) return cb(err)
+      this.setTxStatusSubmitted(txId, rawTx)
```
Comment (kumavis) [inline app/scripts/transaction-manager.js]: dont need to pass this `rawTx` bc we wont emit it

File: app/scripts/transaction-manager.js [hunk 6]
```diff
@@ -144,46 +145,76 @@ module.exports = class TransactionManager extends EventEmitter {
...
+      cb()
+    })
+  }
+
+  validateTxParams (txParams, cb) {
```
Comment (kumavis) [inline app/scripts/transaction-manager.js]: this could be move into tx-utils or provider-engine

File: app/scripts/transaction-manager.js [hunk 7]
```diff
@@ -294,13 +334,16 @@ module.exports = class TransactionManager extends EventEmitter {
   //  should set the status in txData
   //    - `'unapproved'` the user has not responded
   //    - `'rejected'` the user has responded no!
+  //    - `'approved'` the user has approved the tx
   //    - `'signed'` the tx is signed
   //    - `'submitted'` the tx is sent to a server
   //    - `'confirmed'` the tx has been included in a block.
-  _setTxStatus (txId, status) {
+  // "value" is an optional parameter to emit
+  _setTxStatus (txId, status, value) {
```
Comment (kumavis) [inline app/scripts/transaction-manager.js]: dont need value
remove comment too

**Comment by kumavis** [comment]
this needs QA
actual tx signing in ui is broken

**Comment by Zanibas** [comment]
Gas calculation is broken here, because getMaxTxAndFee isn't ever called, per @frankiebee

**Comment by frankiebee** [comment]
working on it

**Comment by frankiebee** [comment]
woops one more thing

**Comment by frankiebee** [comment]
Fixed tx rendering of values Should be good too go!

**Comment by kumavis** [comment]
we may be returning the tx hash in the wrong format

**Comment by kumavis** [comment]
ok yeah we are not returning the txHash - im on it

**Comment by kumavis** [comment]
these commits should fix:
* a tx submission bug (returned raw instead of hash)
* metamask badge sometimes shows wrong number (was including txs from different network)
* metamask popup doesnt show balances until next block

**Review by frankiebee — APPROVED** [review]

**Pull request merged**